data_processing/radiology/refined_table/annotation/generate.py

```diff
@@ -0,0 +1,259 @@
+"""
+Generates Refined PNG table for dicom slices in the series that
+
+1) have 3D segmentations
+2) match user's SQL where clause (filter based on dicom metadata)
+
+This process uses dicom and mhd annotation proxy tables.
+
+The PNG table contains paths to dicom png and overlay image that combines dicom and its corresponding segmentation.
+"""
+
+import os, time, shutil
+import click
+import numpy as np
+from PIL import Image 
+from medpy.io import load
+from filehash import FileHash
+
+from data_processing.common.CodeTimer import CodeTimer
+from data_processing.common.Neo4jConnection import Neo4jConnection
+from data_processing.common.config import ConfigSet
+from data_processing.common.sparksession import SparkConfig
+from data_processing.common.custom_logger import init_logger
+import data_processing.common.constants as const
+
+from pyspark.sql import functions as F
+from pyspark.sql.types import StringType, IntegerType, ArrayType, StructType, StructField
+
+logger = init_logger()
+logger.info("Starting data_processing.radiology.refined_table.annotation.generate")
+
+
+def create_dicom_png(src_path, uuid, accession_number, png_dir):
+    """
+    Create a png from src_path image, and save a png in png_dir.
+    """
+    file_path = src_path.split(':')[-1]
+
+    data, header = load(file_path)
+
+    # Convert 2d image to float to avoid overflow or underflow losses.
+    image_2d = data[:,:,0].astype(float)
+
+    # Rescaling grey scale between 0-255
+    image_2d_scaled = (np.maximum(image_2d, 0) / image_2d.max()) * 255.0
+
+    # Convert to uint
+    image_2d_scaled = np.uint8(image_2d_scaled)
+
+    # Save png named as hash of dicom/seg
+    png_path = os.path.join(png_dir, accession_number)
+    os.makedirs(png_path, exist_ok=True)
+
+    png_path = os.path.join(png_path, uuid.split("-")[-1] + '.png')
+
+    im = Image.fromarray(image_2d_scaled)
+    im.save(png_path)
+    
+    return png_path
+   
+
+def create_seg_png(src_path, uuid, accession_number, png_dir):
+    """
+    Create pngs from src_path image, and save pngs in png_dir.
+    Returns an array of (instance_number, seg_png_path) tuple.
+    """
+    # Save png named as hash of dicom/seg
+    png_dir = os.path.join(png_dir, accession_number)
+    os.makedirs(png_dir, exist_ok=True)
+
+    file_path = src_path.split(':')[-1]
+    data, header = load(file_path)
+
+    num_images = data.shape[2]
+
+    # Find the annotated slices with 3d segmentation.
+    # Some reverse engineering.. save the instance numbers 
+    # from the series to identify the dicom slices that were annotated.
+    slices = []
+    for i in range(num_images):
```

Comment:
In the future, if we can settle for having binaries in the delta table and not on the file system, we can apply spark map-reduce parallelization to speed up this for-loop.